core/src/test/java/org/elasticsearch/index/replication/ESIndexLevelReplicationTestCase.java

```diff
@@ -0,0 +1,478 @@
+
+            @Override
+            public long throttleTimeInNanos() {
+                return 0;
+            }
+        };
+        return new Store(shardId, indexSettings, directoryService, new DummyShardLock(shardId));
+    }
+
+    protected ReplicationGroup createGroup(int replicas) throws IOException {
+        final Path homePath = createTempDir();
+        Settings build = Settings.builder().put(IndexMetaData.SETTING_VERSION_CREATED, Version.CURRENT)
+            .put(IndexMetaData.SETTING_NUMBER_OF_REPLICAS, replicas)
+            .put(IndexMetaData.SETTING_NUMBER_OF_SHARDS, 1)
+            .build();
+        IndexMetaData metaData = IndexMetaData.builder(index.getName()).settings(build).primaryTerm(0, 1).build();
+        return new ReplicationGroup(metaData, homePath);
+    }
+
+    private DiscoveryNode getDiscoveryNode(String id) {
+        return new DiscoveryNode(id, id, DummyTransportAddress.INSTANCE, Collections.emptyMap(),
+            Collections.singleton(DiscoveryNode.Role.DATA), Version.CURRENT);
+    }
+
+    private IndexShard newShard(boolean primary, DiscoveryNode node, IndexMetaData indexMetaData, Path homePath) throws IOException {
+        // add node name to settings for propper logging
+        final Settings nodeSettings = Settings.builder().put("node.name", node.getName()).build();
+        final IndexSettings indexSettings = new IndexSettings(indexMetaData, nodeSettings);
+        ShardRouting shardRouting = TestShardRouting.newShardRouting(shardId, node.getId(), primary, ShardRoutingState.INITIALIZING);
+        final Path path = Files.createDirectories(homePath.resolve(node.getId()));
+        final NodeEnvironment.NodePath nodePath = new NodeEnvironment.NodePath(path);
+        ShardPath shardPath = new ShardPath(false, nodePath.resolve(shardId), nodePath.resolve(shardId), shardId);
+        Store store = createStore(indexSettings, shardPath);
+        IndexCache indexCache = new IndexCache(indexSettings, new DisabledQueryCache(indexSettings), null);
+        MapperService mapperService = MapperTestUtils.newMapperService(homePath, indexSettings.getSettings());
+        for (Map.Entry<String, String> type : indexMapping.entrySet()) {
+            mapperService.merge(type.getKey(), new CompressedXContent(type.getValue()), MapperService.MergeReason.MAPPING_RECOVERY, true);
+        }
+        SimilarityService similarityService = new SimilarityService(indexSettings, Collections.emptyMap());
+        final IndexEventListener indexEventListener = new IndexEventListener() {
+        };
+        final Engine.Warmer warmer = searcher -> {
+        };
+        return new IndexShard(shardRouting, indexSettings, shardPath, store, indexCache, mapperService, similarityService, null, null,
+            indexEventListener, null, threadPool, BigArrays.NON_RECYCLING_INSTANCE, warmer, Collections.emptyList(),
+            Collections.emptyList());
+    }
+
+
+    class ReplicationGroup implements AutoCloseable {
```

Comment:
maybe just implement `Iterable<IndexShard>` here and inlcude the primary... would make the streams below simpler and we can simply write:

```Java

for (IndexShard s : this) {
  s.refresh() | s.flush();
}

```